File: C++/Single_Number_III.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        if (nums.size() < 2) return vector<int>();
        vector<int>ret(2,0);
        int xr = 0, tmp = 0;
        /* first getting the xor of the two numbers */
        for (int x : nums) { xr = xr ^ x; }
        /*Two numbers differ in at least one bit, 
        Assume a has bit k=1, b has bit k=0 
        then classify this array into two sub-arrays, 
        one has all numbers whose bit k=1,
        the other has all numbers whose bit k=0 */
        int mask = 1;
        while ((xr & mask) == 0) mask <<= 1;
        for (int x : nums) {
            if (x & mask) ret[1] ^= x;
        }
        ret[0] = ret[1]^xr;
        return ret;
    }
};

File: Go/Single_Number_III.go
func singleNumber(nums []int) []int {
    ret := []int{0, 0}
    xor := 0
    for i := range(nums) {
        xor ^= nums[i]
    }
    mask := 1
    // find a bit which only one of them has, let this number be a
    for mask & xor == 0 {
        mask <<= 1
    }
    // find all numbers share that bit with a, xor all of them
    // then get a
    for i := range(nums) {
        if mask & nums[i] == mask {
            ret[0] ^= nums[i]
        }
    }
    ret[1] = ret[0] ^ xor
    return ret
}
